feat: use packagr to build as angular package format

terencehonles opened the pull request:
This PR attempts to convert this repo to use ng-packagr to produce builds according to the Angular package format which is needed for AoT builds, and will probably be required as we progress from Angular 7 to beyond.

I tried to leave most things as they were, but the changes I made were the following:
- Fix imports to be relative within a package, and absolute across packages
- Simplify how the package.json is templated (only the package.json keys defined in `scripts/update-package-json.js` are copied from the root into the subpackage package.json)

This does seem to work for us, but there are a couple things that have not been tested or are things to point out:
- ng-packagr does not load in peer libraries, so in order to allow the build process to be automated I ~install~ copy the core library to node_modules after building it and then build the peer libraries. I believe this is related to the issue ng-packagr/ng-packagr#1174 and there is not a more elegant solution to this. (If there is then that should be done)
- because of the above point I'm not sure how easy the development flow is, but building from scratch does not seem to take _too_ long.
- I only tested this when building the package with the Angular compiler. I do not know if the the other bundles are working, and running some tests would help :)

doom777 commented:
hey, can you forward this pull request also to https://github.com/agm-preview/angular-google-maps

doom777 commented:
hey @terencehonles . Thanks for sending the PR. Can you please merge the changes from master into your branch? Make sure it's FF only so there are no merge commits.

Thanks in advance!

terencehonles commented:
@doom777 @SebastianM this PR should be updated, but as I noted in the description I only did limited testing in an Angular 7/8 app.

doom777 commented on package.json:
```diff
@@ -16,28 +16,30 @@
-    "clean": "rimraf dist && rimraf .tmp && mkdir dist && mkdir .tmp && rimraf 'packages/**/*.ngfactory.ts'",
-    "prebuild": "npm run clean",
-    "build": "npm run lint && npm run scripts && npm run packagejson && npm run copyassets",
+    "clean": "rimraf dist rimraf node_modules/@agm rimraf .tmp 'packages/**/*.ngfactory.ts' && mkdirp dist .tmp",
```
shouldn't there be an `&&` between dist and rimraf, otherwise you're trying to delete the rimraf directory.

terencehonles replied:
I'm not sure why I left `rimraf` in there. `rimraf` can take multiple arguments and I didn't want any `&&`'s but I didn't need to leave the extra `rimraf`'s in there

terencehonles commented:
I updated the code to match master. I haven't had a chance to test it, but it does build w/o any warnings.

doom777 commented on packages/core/package.json:
```diff
@@ -0,0 +1,26 @@
+    "angular2"
+  ],
+  "peerDependencies": {
+    "@angular/common": "^6.0.0 || ^7.0.0",
```
we now support angular 8

doom777 commented on packages/js-marker-clusterer/package.json:
```diff
@@ -1,6 +1,29 @@
+  ],
+  "peerDependencies": {
+    "@angular/core": "^6.0.0 || ^7.0.0 || ^8.0.0",
+    "@agm/core": "^1.0.0-beta.6",
```
agm 7

doom777 commented on packages/snazzy-info-window/package.json:
```diff
@@ -15,15 +11,19 @@
-  },
   "peerDependencies": {
     "@angular/core": "^6.0.0 || ^7.0.0 || ^8.0.0",
     "@agm/core": "^1.0.0-beta.6",
```
7 here also

terencehonles commented:
I updated some more of the dev dependencies and cleaned up some of the tools I made unnecessary and fixed a test that broke when upgrading versions (I had also fixed it on #1648 ).

I still haven't tested it aside from ensuring it builds, but it should be up to date now.

terencehonles commented:
It does seem the changes still build against the application we have, but I don't have the right data locally to run it so I'll have to test running the actual application elsewhere later.

terencehonles commented:
The changes do still seem to work as before. I don't any package other than `@agm/core` and I only use angular w/ aot so if anyone has a different workflow or uses the other packages they should test those.

terencehonles commented:
Let's not let this get stale again. It wasn't my fault this wasn't merged in earlier. There was some clean up I was able to add because there was more time, but this PR is basically the same as it was in May aside from resolving merge conflicts.

We need to get this into mainline so changes are made with respect to this change rather than me trying to resolve what happened in the meanwhile.

If there is concern about the other builds, we need to add some tests or easy way of verifying they are built correctly. The existing bundle formats that have been moved/updated should be prioritized, and new formats should be fixed as they are noticed to be broken (if they even are)

Just my 2¢

doom777 commented:
let's give @SebastianM a few days to review it.

doom777 commented on package.json:
```diff
@@ -16,28 +16,30 @@
+    "clean": "rimraf dist node_modules/@agm 'packages/**/*.ngfactory.ts' && mkdirp dist",
+    "prebuild": "yarn clean",
+    "build": "yarn lint && yarn build:all",
+    "build:all": "yarn build:core && yarn packagejson && mkdirp node_modules/@agm/core && ncp dist/@agm/core node_modules/@agm/core && yarn build:js-marker-clusterer && yarn build:snazzy-info-window && yarn packagejson",
```
why does `yarn packagejson` show up twice?

terencehonles replied:
I'm guessing you figured it out, but it makes sure the `scripts/update-package-json.js` gets run on `@agm/core` which gives it a version number and some other `package.json` data, which allows the compiler to not complain when building the dependent packages.

This might actually have been from when I was using `npm` to install `@agm/core` and _may_ not be needed if nothing checks its `package.json` while building the dependent packages.
